pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -189,25 +190,76 @@ func injectApmTelemetryConfig(pod *corev1.Pod) {
+	}
+
+	// Tracing libraries for language detection
+	libsToInject = getLibrariesLanguageDetection(pod, registry)
+	if len(libsToInject) > 0 {
+		autoDetected = true
+		return libsToInject, autoDetected
+	}
+
+	// Latest tracing libraries for all supported languages (java, js, dotnet, python, ruby)
+	libsToInject = getAllLatestLibraries(registry)
+
+	return libsToInject, autoDetected
+}
+
+// getPinnedLibraries returns tracing libraries to inject as configured by apm_config.instrumentation.lib_versions
+func getPinnedLibraries(registry string) []libInfo {
+	libsToInject := []libInfo{}
 	var libVersion string
-	singleStepLibraryVersions := config.Datadog.GetStringMapString("apm_config.instrumentation.lib_versions")
 
-	for _, lang := range supportedLanguages {
-		libVersion = "latest"
+	singleStepLibraryVersions := config.Datadog.GetStringMapString("apm_config.instrumentation.lib_versions")
 
-		if version, ok := singleStepLibraryVersions[string(lang)]; ok {
-			log.Warnf("Library version %s is specified for language %s", version, string(lang))
+	// If APM Instrumentation is enabled and configuration apm_config.instrumentation.lib_versions specified, inject only the libraries from the configuration
+	for lang, version := range singleStepLibraryVersions {
+		if !slices.Contains(supportedLanguages, language(lang)) {
+			log.Warnf("APM Instrumentation detected configuration for unsupported language: %s. Tracing library for %s will not be injected", lang, lang)
+		} else {
+			log.Infof("Library version %s is specified for language %s", version, lang)
 			libVersion = version
+			libsToInject = append(libsToInject, libInfo{lang: language(lang), image: libImageName(registry, language(lang), libVersion)})
 		}
+	}
+	return libsToInject
+}
 
-		libsToInject[string(lang)] = libInfo{
-			lang:  lang,
-			image: libImageName(registry, lang, libVersion),
+// getLibrariesLanguageDetection runs process language auto-detection and returns languages to inject for APM Instrumentation.
+// The langages information is available in workloadmeta-store and attached on the pod's owner.
+func getLibrariesLanguageDetection(pod *corev1.Pod, registry string) []libInfo {
+	if config.Datadog.GetBool("admission_controller.run_language_detection") {
+		// Always run language detection to collect telemetry data
+		libsToInject := getAutoDetectedLibraries(pod, registry)
+		if config.Datadog.GetBool("admission_controller.inject_auto_detected_libraries") {
+			// Use libraries returned by language detection for APM Instrumentation
+			return libsToInject
 		}
 	}
-	return libsToInject
+
+	return []libInfo{}
+}
+
+// getAllLatestLibraries returns all supported by APM Instrumentation tracing libraries
+func getAllLatestLibraries(registry string) []libInfo {
+	libsToInject := []libInfo{}
+
+	for _, lang := range supportedLanguages {
+		libsToInject = append(libsToInject, libInfo{lang: language(lang), image: libImageName(registry, lang, "latest")})
+	}
+
+	return []libInfo{}
```

Comment:
libsToInject ?

Reply:
Thanks. Didn't notice it failed unit-tests.